Add whitelist of safe intrinsics

frazar opened the pull request:
This PR should fix #5996, where intrinsic operations where all marked as unsafe.

I'm rather new to this codebase, so I might be doing something *very* wrong. Please forgive me!

In particular, I'm not sure how to "check that we are in extern `rust-intrinsics`" as mentioned [in this comment](https://github.com/rust-analyzer/rust-analyzer/issues/5996#issuecomment-709234802).

lnicola commented:
Awesome :-). In other places we clone the repository and read the files from there, but this feels like such a huge wart TBH...

lnicola commented on crates/hir_expand/src/name.rs:
```diff
@@ -208,6 +208,42 @@ pub mod known {
         PartialOrd,
         Eq,
         PartialEq,
+        // Safe primitives
```
```suggestion
        // Safe intrinsics
```

lnicola commented on crates/hir_def/src/item_tree/tests.rs:
```diff
@@ -437,3 +437,23 @@ fn assoc_item_macros() {
     );
 }
+
+#[test]
```
Can you also add a test (or change this one) with an unsafe intrinsic?

frazar replied:
I moved the tests in a better location, and included both a safe and an unsafe intrinsic

lnicola commented:
Regarding the type of the extern block, the AST looks like this:

```
  EXTERN_BLOCK@0..28
    ABI@0..10
      EXTERN_KW@0..6 "extern"
      WHITESPACE@6..7 " "
      STRING@7..10 "\"C\""
```

lnicola commented:
I think you can get the `ABI` node using `extern_block.abi()`, but there's no child for the type, only for the `extern` token. @matklad, we should probably add it to the AST?

matklad commented on crates/hir_def/src/item_tree/lower.rs:
```diff
@@ -42,6 +42,45 @@ pub(super) struct Ctx {
+/// Returns `true` if the given intrinsic is unsafe to call or not.
+pub fn is_intrinsic_fn_unsafe(name: &Name) -> bool {
+    // Should be kept in sync with https://github.com/rust-lang/rust/blob/c6e4db620a7d2f569f11dcab627430921ea8aacf/compiler/rustc_typeck/src/check/intrinsic.rs#L68
+    *name != known::abort
```
Let's write this as `[known::a, know::b, ...], contains(&name)`

matklad replied:
Other than that, lgmt!

I think it's fine to punt on attribute check

bors d+

matklad replied:
Oh, and the fn probably can be private? If it is, move it to the bottom of the file, as per https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/style.md#order-of-items

matklad commented:
@lnicola good question! Turns out this needs a bit of design, I've put together https://github.com/rust-analyzer/rust-analyzer/issues/6308

frazar commented:
Thank you for your feedback and comments. I think I addressed all points, except for the issue about checking the ABI name, which I guess is blocked on #6308.

Also, I'm not sure why CI failed for macos-latest.. It might be unrelated.

Let me know what should I do next!

lnicola commented:
bors r=frazar

frazar commented:
Great! As a final note, would you mind adding the `hacktoberfest-accepted` label to this PR? Thanks!

lnicola commented:
Sure, done.
